# Sprint 6, T01

## Nivell 1

### Execise 1 - samplig of some database with sport theme, sample random simple and sample sistematic

Let's analyze a dataset about earnings in esports - https://www.kaggle.com/jackdaoud/esports-earnings-for-players-teams-by-game. I will work with the amount of earnings from each player to represent the statistics about the highest earnings. I have 1000 total observations from the top 10 games on eSports.

In [4]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import random
import statistics

In [ ]:
%matplotlib inline
from matlab import random
from random import randint

In [5]:
esports_df = pd.read_csv('highest_earning_players.csv')

In [6]:
esports_df.head(5)

PlayerId NameFirst   NameLast CurrentHandle CountryCode  TotalUSDPrize  \
0      3883     Peter  Rasmussen       dupreeh          dk     1822989.41   
1      3679   Andreas   Højsleth         Xyp9x          dk     1799288.57   
2      3885   Nicolai     Reedtz        dev1ce          dk     1787489.88   
3      3672     Lukas  Rossander        gla1ve          dk     1652350.75   
4     17800      Emil       Reif        Magisk          dk     1416448.64   

                               Game                 Genre  
0  Counter-Strike: Global Offensive  First-Person Shooter  
1  Counter-Strike: Global Offensive  First-Person Shooter  
2  Counter-Strike: Global Offensive  First-Person Shooter  
3  Counter-Strike: Global Offensive  First-Person Shooter  
4  Counter-Strike: Global Offensive  First-Person Shooter

In [7]:
esports_df.describe()

PlayerId  TotalUSDPrize
count   1000.000000   1.000000e+03
mean   27793.587000   3.977932e+05
std    22170.225194   6.908492e+05
min     1000.000000   2.417167e+04
25%     5374.500000   8.378962e+04
50%    23502.000000   1.683284e+05
75%    48127.250000   3.937352e+05
max    83085.000000   6.952597e+06

#### Simple random sample

First I will generate a simple random sample with n observations.

In [7]:
simple_esports_df = esports_df.sample(n = 100)
simple_esports_df.head(5)

PlayerId   NameFirst NameLast CurrentHandle CountryCode  TotalUSDPrize  \
252     41044  Tian Liang      Gao          Tian          cn      268282.03   
498     28443       Scott  Kennedy         Custa          au       53759.68   
302     25047    Harrison    Chang         psalm          us     1873138.80   
46      11483       Peter   Jarguz     stanislaw          ca      481487.38   
356     54842        Nick  McGuire        Aspect          us      215391.67   

                                 Game                            Genre  
252                 League of Legends  Multiplayer Online Battle Arena  
498                         Overwatch             First-Person Shooter  
302                          Fortnite                    Battle Royale  
46   Counter-Strike: Global Offensive             First-Person Shooter  
356                          Fortnite                    Battle Royale

In [23]:
simple_esports_df.info()#I want to check if I really have 100 values

<class 'pandas.core.frame.DataFrame'>
Int64Index: 100 entries, 252 to 15
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   PlayerId       100 non-null    int64  
 1   NameFirst      100 non-null    object 
 2   NameLast       100 non-null    object 
 3   CurrentHandle  100 non-null    object 
 4   CountryCode    100 non-null    object 
 5   TotalUSDPrize  100 non-null    float64
 6   Game           100 non-null    object 
 7   Genre          100 non-null    object 
dtypes: float64(1), int64(1), object(6)
memory usage: 7.0+ KB


In [8]:
simple_esports_df.describe()

PlayerId  TotalUSDPrize
count    100.000000   1.000000e+02
mean   27141.170000   4.004989e+05
std    21457.334855   5.135877e+05
min     1049.000000   2.967600e+04
25%     8521.750000   8.297275e+04
50%    21963.000000   1.654695e+05
75%    41733.250000   5.519888e+05
max    75020.000000   2.995620e+06

Interpretation: The arithmetic mean is 400498.9 USD prize by player. For a very simple statistics it might be ok but we need to get average rate and correctly weight each data point. Let's use the harmonic mean.

In [143]:
har_mean1 = statistics.harmonic_mean(simple_esports_df['TotalUSDPrize'])
har_mean1

130079.46524286638

Interpretation: The harmonic mean is 130079.47 USD prize by player. It seems more fair conclusion for the average prize by player.

#### Sistematic sample

We will need a random number and step number using modulation and functions.

In [39]:
step_num = 10

def randomNum():
    return np.random.randint(1, 50)

In [40]:
#https://sparkbyexamples.com/pandas/pandas-select-dataframe-rows-by-index/
#[start:stop:step]
systematic_esports_df = esports_df[num:len(esports_df):step_num]
systematic_esports_df.head(5)

PlayerId NameFirst   NameLast CurrentHandle CountryCode  TotalUSDPrize  \
38      4997     Paweł  Bieliński         byali          pl      574103.56   
48     11938   Timothy         Ta     autimatic          us      449799.56   
58      2456   Lincoln        Lau           fnx          br      387551.81   
68      5499    Miikka     Kemppi         suNny          fi      326080.00   
78     39608    Alexey    Golubev        Qikert          kz      275404.13   

                                Game                 Genre  
38  Counter-Strike: Global Offensive  First-Person Shooter  
48  Counter-Strike: Global Offensive  First-Person Shooter  
58  Counter-Strike: Global Offensive  First-Person Shooter  
68  Counter-Strike: Global Offensive  First-Person Shooter  
78  Counter-Strike: Global Offensive  First-Person Shooter

In [41]:
systematic_esports_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 97 entries, 38 to 998
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   PlayerId       97 non-null     int64  
 1   NameFirst      97 non-null     object 
 2   NameLast       97 non-null     object 
 3   CurrentHandle  97 non-null     object 
 4   CountryCode    97 non-null     object 
 5   TotalUSDPrize  97 non-null     float64
 6   Game           97 non-null     object 
 7   Genre          97 non-null     object 
dtypes: float64(1), int64(1), object(6)
memory usage: 6.2+ KB


I have only 96 entries so I would like to add the rest to get 100 entries with the same step.

In [42]:
len(systematic_esports_df)

97

In [43]:
addNum = (100 - len(systematic_esports_df))
addNum

3

In [44]:
x = step_num - addNum
var1 = esports_df.iloc[x]
var2 = esports_df.iloc[x+step_num]
var3 = esports_df.iloc[x+(2*step_num)]

In [45]:
systematic_esports_df = systematic_esports_df.append(var1, ignore_index = True)
systematic_esports_df = systematic_esports_df.append(var2, ignore_index = True)
systematic_esports_df = systematic_esports_df.append(var3, ignore_index = True)

In [46]:
systematic_esports_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   PlayerId       100 non-null    int64  
 1   NameFirst      100 non-null    object 
 2   NameLast       100 non-null    object 
 3   CurrentHandle  100 non-null    object 
 4   CountryCode    100 non-null    object 
 5   TotalUSDPrize  100 non-null    float64
 6   Game           100 non-null    object 
 7   Genre          100 non-null    object 
dtypes: float64(1), int64(1), object(6)
memory usage: 6.4+ KB


Now I have 100 entries and I can compare harmonic mean of two methods of sampling.

In [47]:
har_mean2 = statistics.harmonic_mean(systematic_esports_df['TotalUSDPrize'])
har_mean2

117705.22126829364

Interpretation: The harmonic mean is 117705.22 USD prize by player. That means that both methods give different results.

## Nivell 2

### Execise 2 - stratified random sample and SMOTE

#### Stratified random sample

For stratified random sample I will divide the palayer in groups by game, there are 10 games in the dataset, so I can pick 10 players form each(10%) and than I will get 100 players in total.

In [44]:
esports_df['Game'].value_counts()

Heroes of the Storm                 100
Fortnite                            100
Arena of Valor                      100
Starcraft II                        100
Dota 2                              100
Counter-Strike: Global Offensive    100
PUBG                                100
Hearthstone                         100
League of Legends                   100
Overwatch                           100
Name: Game, dtype: int64

In [55]:
sorted_esports_df = esports_df. sort_values(by='Game')

In [57]:
sorted_esports_df.head(5)

PlayerId NameFirst       NameLast CurrentHandle CountryCode  \
801     60565    Guixin            Guo        NuoYan          cn   
873     75020       Aum  Chatchanapong         Happy          th   
872     54012   Denchai     Eampaojeen        JameCo          th   
871     53974         -              -       Sirenia          tw   
870     53973         -              -         Liang          tw   

     TotalUSDPrize            Game                            Genre  
801      416162.04  Arena of Valor  Multiplayer Online Battle Arena  
873       40376.87  Arena of Valor  Multiplayer Online Battle Arena  
872       41144.36  Arena of Valor  Multiplayer Online Battle Arena  
871       42482.43  Arena of Valor  Multiplayer Online Battle Arena  
870       42482.43  Arena of Valor  Multiplayer Online Battle Arena

In [68]:
sorted_esports_df['Game'].unique()

array(['Arena of Valor', 'Counter-Strike: Global Offensive', 'Dota 2',
       'Fortnite', 'Hearthstone', 'Heroes of the Storm',
       'League of Legends', 'Overwatch', 'PUBG', 'Starcraft II'],
      dtype=object)

In [67]:
group1 = sorted_esports_df[sorted_esports_df['Game'] == 'Arena of Valor']
group1.info()#to check if I get 100 entruies for this group, than I will apply the code for all 10 groups

<class 'pandas.core.frame.DataFrame'>
Int64Index: 100 entries, 801 to 835
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   PlayerId       100 non-null    int64  
 1   NameFirst      100 non-null    object 
 2   NameLast       100 non-null    object 
 3   CurrentHandle  100 non-null    object 
 4   CountryCode    100 non-null    object 
 5   TotalUSDPrize  100 non-null    float64
 6   Game           100 non-null    object 
 7   Genre          100 non-null    object 
dtypes: float64(1), int64(1), object(6)
memory usage: 7.0+ KB


In [69]:
group2=sorted_esports_df[sorted_esports_df['Game'] == 'Counter-Strike: Global Offensive'] 
group3=sorted_esports_df[sorted_esports_df['Game'] == 'Dota 2'] 
group4=sorted_esports_df[sorted_esports_df['Game'] == 'Fortnite'] 
group5=sorted_esports_df[sorted_esports_df['Game'] == 'Hearthstone'] 
group6=sorted_esports_df[sorted_esports_df['Game'] == 'Heroes of the Storm'] 
group7=sorted_esports_df[sorted_esports_df['Game'] == 'League of Legends'] 
group8=sorted_esports_df[sorted_esports_df['Game'] == 'Overwatch'] 
group9=sorted_esports_df[sorted_esports_df['Game'] == 'PUBG'] 
group10=sorted_esports_df[sorted_esports_df['Game'] == 'Starcraft II']

In [70]:
#getting 10 entries from each group
esports_df1 = group1.sample(n = 10)
esports_df2 = group2.sample(n = 10)
esports_df3 = group3.sample(n = 10)
esports_df4 = group4.sample(n = 10)
esports_df5 = group5.sample(n = 10)
esports_df6 = group6.sample(n = 10)
esports_df7 = group7.sample(n = 10)
esports_df8 = group8.sample(n = 10)
esports_df9 = group9.sample(n = 10)
esports_df10 = group10.sample(n = 10)

In [72]:
stratified_esports_df = esports_df1.append((esports_df2,
                                            esports_df3,
                                            esports_df4,
                                            esports_df5,
                                            esports_df6,
                                            esports_df7,
                                            esports_df8,
                                            esports_df9,
                                            esports_df10),ignore_index = True)

In [76]:
stratified_esports_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   PlayerId       100 non-null    int64  
 1   NameFirst      100 non-null    object 
 2   NameLast       100 non-null    object 
 3   CurrentHandle  100 non-null    object 
 4   CountryCode    100 non-null    object 
 5   TotalUSDPrize  100 non-null    float64
 6   Game           100 non-null    object 
 7   Genre          100 non-null    object 
dtypes: float64(1), int64(1), object(6)
memory usage: 6.4+ KB


In [145]:
har_mean3 = statistics.harmonic_mean(stratified_esports_df['TotalUSDPrize'])
har_mean3

132956.6247256477

Interpretation: The harmonic mean is 132956.62 USD prize by player. That means that by applying stratifies sampling methos I got similar value to simple sampling mathod.

#### SMOTE

In [80]:
import imblearn
from imblearn.over_sampling import SMOTE

In [82]:
esports_df['CountryCode'].unique()

array(['cn', 'tw', 'vn', 'th', 'kr', 'us', 'my', 'br', 'be', 'es', 'de',
       'tr', 'pl', 'se', 'gb', 'ca', 'sk', 'fr', 'kz', 'ru', 'fi', 'bg',
       'uz', 'dk', 'cz', 'no', 'ee', 'nl', 'ua', 'ba', 'ph', 'mk', 'mo',
       'jo', 'sg', 'au', 'by', 'ro', 'il', 'pk', 'lb', 'at', 'hk', 'ar',
       'ch', 'lv', 'si', 'lt', 'jp', 'gr', 'it', 'hr', 'cu', 'gi', 'ie',
       'mx'], dtype=object)

In [89]:
esports_df.groupby(['CountryCode']).size()#how many player I have by country? I will pick two countries for SMOTE

CountryCode
ar      3
at      3
au      5
ba      1
be      5
bg      6
br     10
by      2
ca     37
ch      3
cn    168
cu      1
cz      3
de     30
dk     36
ee      2
es      6
fi     16
fr     37
gb     16
gi      1
gr      1
hk      5
hr      2
ie      1
il      2
it      2
jo      2
jp      4
kr    251
kz      6
lb      1
lt      2
lv      1
mk      1
mo      1
mx      1
my      6
nl     10
no     10
ph      7
pk      1
pl     14
ro      2
ru     22
se     45
sg      1
si      4
sk      3
th     29
tr      4
tw     34
ua     12
us    114
uz      1
vn      7
dtype: int64

I will take players from South Korea - KR and China - CN because these countries has the highest number of players.

In [95]:
groupKR=esports_df[esports_df['CountryCode'] == 'kr']
groupCN=esports_df[esports_df['CountryCode'] == 'cn']

In [96]:
smote_esports_df = groupKR.append(groupCN,ignore_index = True)

In [98]:
smote_esports_df.head(5)

PlayerId   NameFirst NameLast CurrentHandle CountryCode  TotalUSDPrize  \
0     53998  Chang Hoon     Shin          JJak          kr      122650.01   
1     53996     Ho Yeon      Lee          Rush          kr      117894.70   
2     53999     Do Yeop      Kim           HAK          kr      111710.36   
3     53995     Sun Woo      Kim           Sun          kr      111582.45   
4     53997   Hyung Min      Kim        Chaser          kr       71604.67   

             Game                            Genre  
0  Arena of Valor  Multiplayer Online Battle Arena  
1  Arena of Valor  Multiplayer Online Battle Arena  
2  Arena of Valor  Multiplayer Online Battle Arena  
3  Arena of Valor  Multiplayer Online Battle Arena  
4  Arena of Valor  Multiplayer Online Battle Arena

In [130]:
mod_smote_esports_df = smote_esports_df[['CountryCode','TotalUSDPrize']]
mod_smote_esports_df.head()

CountryCode  TotalUSDPrize
0          kr      122650.01
1          kr      117894.70
2          kr      111710.36
3          kr      111582.45
4          kr       71604.67

In [131]:
pd.options.mode.chained_assignment = None
mod_smote_esports_df['CountryCode'] = np.where((mod_smote_esports_df['CountryCode'] == 'kr'), 1, 0)

In [132]:
mod_smote_esports_df

CountryCode  TotalUSDPrize
0              1      122650.01
1              1      117894.70
2              1      111710.36
3              1      111582.45
4              1       71604.67
..           ...            ...
414            0      125173.05
415            0      142369.56
416            0      184811.62
417            0      134300.51
418            0      172891.85

[419 rows x 2 columns]

In [133]:
mod_smote_esports_df['CountryCode'].value_counts()

1    251
0    168
Name: CountryCode, dtype: int64

That means that I have 251 player from South Korea and 168 player from China. With SMOTE I will increase the values of China player to 251.

In [134]:
smote = SMOTE(sampling_strategy = 'minority')
X_sm, y_sm = smote.fit_resample(mod_smote_esports_df, mod_smote_esports_df['CountryCode'])

In [136]:
X_sm['CountryCode'].value_counts()

0    251
1    251
Name: CountryCode, dtype: int64

In [146]:
har_mean4 = statistics.harmonic_mean(X_sm['TotalUSDPrize'])
har_mean4

114032.44263264671

Interpretation: the average prize of Soouth Korea and China players is lower than the average prize of player from all countries.

## Nivell 3

### Execise 3 - Reservoir samplig

I'm using the strategy from https://towardsdatascience.com/the-5-sampling-algorithms-every-data-scientist-need-to-know-43c7bc11d17c

In [49]:
def generator(max):
    number = 1
    while number < max:
        number += 1
        yield number
        
stream = esports_df['TotalUSDPrize']

k=100 #number of samples
reservoir = []#list of values
for i, element in enumerate(stream):
    if i+1<= k:
        reservoir.append(element)
    else:
        probability = k/(i+1)
        if random.random() < probability:
            # select item in stream and remove one of the k items already selected
            reservoir[random.choice(range(0,k))] = element
print(reservoir)

[89725.4, 133558.02, 1778125.61, 1652350.75, 1416448.64, 47973.61, 74885.91, 172583.33, 159650.52, 591271.39, 250580.01, 964635.39, 85535.96, 296734.75, 112482.56, 298622.85, 933853.71, 106288.65, 835376.43, 58710.01, 164025.25, 133917.25, 106462.18, 118657.27, 164396.18, 538926.12, 360841.67, 297730.97, 193787.59, 360093.52, 205477.15, 82032.82, 64140.31, 70255.34, 153116.49, 38892.33, 94813.8, 77723.5, 2864004.17, 1882492.98, 414258.4, 65922.73, 310008.33, 111579.95, 1272450.74, 54159.21, 2007841.6, 228500.0, 449799.56, 1873138.8, 51663.14, 171767.15, 417465.86, 204049.56, 138012.28, 106225.0, 168959.28, 488379.08, 206587.79, 89529.34, 1109880.47, 293345.18, 437975.0, 616412.36, 89214.8, 135217.79, 86588.9, 74175.89, 150900.0, 58295.94, 1202225.88, 63634.48, 775574.81, 134800.0, 347699.25, 61228.35, 284845.29, 165169.85, 189265.81, 56726.85, 167442.47, 1081544.74, 2551657.34, 131670.38, 156959.11, 84196.0, 217500.0, 90439.44, 615626.39, 1126345.35, 75125.32, 65922.73, 153868.98, 6952

In [140]:
len(reservoir)

100

In [50]:
a = min(reservoir)
b = max(reservoir)
print(a, b)

38892.33 6952596.58


In [147]:
har_mean5 = statistics.harmonic_mean(reservoir)
har_mean5

120491.53536710396

Interpretation: the average prize 120491.53 might be the most correct answer because it's close to the value from simple sampling and pointing toward higher values seen with stratified method.

In [149]:
print('Average prize by simple sampling is', har_mean1)
print('Average prize by sistemating sampling is', har_mean2)
print('Average prize by stratified sampling is', har_mean3)
print('Average prize by SMOTE sampling is', har_mean4)
print('Average prize by reservoir sampling is', har_mean5)

Average prize by simple sampling is 130079.46524286638
Average prize by sistemating sampling is 119455.27635236355
Average prize by stratified sampling is 132956.6247256477
Average prize by SMOTE sampling is 114032.44263264671
Average prize by reservoir sampling is 120491.53536710396
